File: backend/maze.py
#!/usr/bin/python

from IRobot import IRobot

class Point(object):
    def __init__(self, x, y):
        self.x = x
        self.y = y
    def __str__(self):
        return str((self.x, self.y))
    def __repr__(self):
        return str((self.x, self.y))
    def __eq__(self,p):
        return (self.x == p.x and self.y == p.y)
    def __ne__(self,p):
        return not(self.x == p.x and self.y == p.y)

class Maze(object):
    def __init__(self, mazeCol, rows=None):
        if rows != None:
            self.maze = [[IRobot.WALL for row in range(rows)] for column in range(mazeCol)]
            self.height = rows
            self.width = mazeCol
        else:
            self.maze = mazeCol
            self.height = len(mazeCol[0])
            self.width = len(mazeCol)
        self.start = Point(1,1)
        self.location = Point(1,1)
        self.target = Point(self.width - 2, self.height - 2)
        self.heading = IRobot.EAST


    def __str__(self):
        return str(self.maze)+str((self.start, self.target, self.location, self.heading))

    def __repr__(self):
        result = ""
        for row in range(self.height):
            for column in self.maze:
                if column[row] == IRobot.PASSAGE:
                    result += "-"
                else:
                    result += "X"
            result += '\n'
        result += "Start: "+str(self.start)+"\n"
        result += "Target: "+str(self.target)+"\n"
        result += "Location: "+str(self.location)+"\n"  
        result += "Heading: "+str(self.heading)
        return result

    def getWidth(self):
        return self.width

    def getHeight(self):
        return self.height

    def getTileType(self, px, py=None):
        if(py != None):
            return self.maze[px][py]
        else:
            return self.maze[px.x][px.y]

    def setTile(self, tile, px, py=None):
        if(py!=None):
            self.maze[px][py] = tile
        else:
            self.maze[px.x][px.y] = tile

    def getStart(self):
        return self.start
    def setStart(self,px,py=None):
        if (py != None):
            self.start = Point(px,py)
        else:
            self.start = px

    def getLocation(self):
        return self.location
    def setLocation(self,px,py=None):
        if (py!=None):
            self.location = Point(px,py)
        else:
            self.location = px

    def getFinish(self):
        return self.target
    def setFinish(self,px,py=None):
        if (py!=None):
            self.target = Point(px,py)
        else:
            self.target = px

    def getHeading(self):
        return self.heading
    def setHeading(self, newHeading):
        if(newHeading < IRobot.NORTH or newHeading > IRobot.WEST):
            raise RuntimeError("The robot's heading can only be NORTH, SOUTH, EAST or WEST.")
        self.heading = newHeading

if __name__ == "__main__":
    maze = Maze(5,3)
    print(repr(maze))

File: backend/java/src/Maze.java
import java.awt.Point;

public class Maze {
    int[][] maze;
    int height;
    int width;
    Point start;
    Point target;
    Point location;
    int heading;

    public Maze(int col, int row) {
        width = col;
        height = row;
        maze = new int[width][height];
        for (int i = 0; i < width; i++) {
            for (int j = 0; j < height; j++) {
                maze[i][j] = IRobot.WALL;
            }
        }
        start = new Point(1,1);
        location = new Point(1,1);
        target = new Point(width - 2, height - 2);
        heading = IRobot.EAST;
    }
    public Maze(int[][] m) {
        maze = transpose(m);
        width = maze.length;
        height = maze[0].length;
        start = new Point(1,1);
        location = new Point(1,1);
        target = new Point(width - 2, height - 2);
        heading = IRobot.EAST;
    }

    public int getWidth() {
        return width;
    }

    public int getHeight() {
        return height;
    }

    public int getTileType(int x, int y) {
        return maze[x][y];
    }
    public int getTileType(Point p) {
        return maze[p.x][p.y];
    }

    public void setTile(int tile, int x, int y) {
        maze[x][y] = tile;
    }
    public void setTile(int tile, Point p) {
        maze[p.x][p.y] = tile;
    }

    public Point getStart() {
        return start;
    }
    public void setStart(int x, int y) {
        start = new Point(x,y);
    }
    public void setStart(Point p) {
        start = p;
    }

    public Point getLocation() {
        return location;
    }
    public void setLocation(int x, int y) {
        location = new Point(x,y);
    }
    public void setLocation(Point p) {
        location = p;
    }

    public Point getFinish() {
        return target;
    }
    public void setFinish(int x, int y) {
        target = new Point(x,y);
    }
    public void setFinish(Point p) {
        target = p;
    }

    public int getHeading() {
        return heading;
    }
    public void setHeading(int newHeading) {
        if(newHeading < IRobot.NORTH || newHeading > IRobot.WEST) {
            throw new RuntimeException("The robot's heading can only be NORTH, SOUTH, EAST or WEST.");
        }
        heading = newHeading;
    }

    public void printMaze() {
        String result = "";
        for (int j = 0; j < height; j++) {
            for (int i = 0; i < width; i++) {
                switch(maze[i][j]) {
                    case IRobot.WALL:
                        result += "X";
                        break;
                    case IRobot.PASSAGE:
                        result += "-";
                        break;
                    case IRobot.BEENBEFORE:
                        result += "=";
                        break;
                }
            }
            result += "\n";
        }
        result += "Start: "+start+"\n";
        result += "Target: "+target+"\n";
        result += "Location: "+location+"\n";
        result += "Heading: "+heading;
        System.out.println(result);
    }

    // transpose incoming 2D list representation of maze
    // if maze = 123  then '6' should be maze[2][1]
    //           456
    // coming in as [[1,2,3],  would make maze[x][y] refer to row x, col y rather
    //               [4,5,6]]  than col x, row y as intended
    // therefore transpose incoming to [[1,4],  so that maze[x][y] refers to
    //                                  [2,5],  col x and row y
    //                                  [3,6]]
    public int[][] transpose(int[][] arr) {
        int l = arr[0].length;
        int w = arr.length;
        int[][] newArr = new int[l][w];
        for (int i = 0; i < l; i++) {
            for (int j = 0; j < w; j++) {
                newArr[i][j] = arr[j][i];
            }
        }
        return newArr;
    }
}